lib/FetchEnvs.js

```diff
@@ -14,235 +14,189 @@
-          break;
+
+    if (decode) {
+      for (const [key, value] of Object.entries(data)) {
+        data[key] = Buffer.from(value, 'base64').toString();
       }
     }
-    return kubeError === undefined ? (result) : { error: kubeError };
+
+    return {...conf, data};
   }
 
-  async _processEnvFrom(envFrom) {
-    return await Promise.all(envFrom.map(async (element) => {
-      const mapData = clone(element);
-      let optional = objectPath.get(element, 'optional', false);
-      let kubeError;
-      if (objectPath.has(element, 'configMapRef')) {
-        let name = objectPath.get(element, 'configMapRef.name');
-        let namespace = objectPath.get(element, 'configMapRef.namespace', this.namespace);
-        let res;
-        try {
-          res = await this.kubeResourceMeta.request({ uri: `/api/v1/namespaces/${namespace}/configmaps/${name}`, json: true });
-        } catch (e) {
-          log.warn(e);
-          kubeError = e.message;
-        }
+  async #genericKeyRef(conf, valueFrom = 'genericKeyRef', decode = false) {
 
-        let data = objectPath.get(res, 'data');
-        if (kubeError === undefined) kubeError = 'no data returned from processEnvFrom. make sure your data exists in the correct location and is in the expected format.';
+    const ref = conf.valueFrom[valueFrom];
+    const strategy = conf.overrideStrategy;
+    const {
+      name,
+      namespace = this.namespace,
+      kind = KIND_MAP.get(valueFrom),
+      key,
+      matchLabels,
+      apiVersion = 'v1',
+      type
+    } = ref;
 
-        if (data) {
-          mapData.data = data;
-        } else if (!optional) {
-          throw new Error(`envFrom.configMapRef.${name}.data not found: ${kubeError}`);
-        } else {
-          let msg = `envFrom.configMapRef.${name}.data not found: ${kubeError}`;
-          log.warn(msg);
-          this.updateRazeeLogs('warn', { controller: 'FetchEnvs', message: msg });
-        }
-      } else if (objectPath.has(element, 'secretMapRef')) {
-        let name = objectPath.get(element, 'secretMapRef.name');
-        let namespace = objectPath.get(element, 'secretMapRef.namespace', this.namespace);
-        let res;
-        try {
-          res = await this.kubeResourceMeta.request({ uri: `/api/v1/namespaces/${namespace}/secrets/${name}`, json: true });
-        } catch (e) {
-          log.warn(e);
-          kubeError = e.message;
-        }
+    const krm = await this.kubeClass.getKubeResourceMeta(
+      apiVersion,
+      kind,
+      'update'
+    );
 
-        let data = objectPath.get(res, 'data');
-        if (kubeError === undefined) kubeError = 'no data returned from processEnvFrom. make sure your data exists in the correct location and is in the expected format.';
+    const qs = labelSelectors(matchLabels);
+    const uri = krm.uri({namespace, kind, name});
+    const response = await this.api({
+      uri: uri,
+      json: true,
+      qs: qs
+    });
```

Comment:
did we lose the try catch around this, or is it at a different scope that im missing?

Reply:
Is it intended that a template is all or nothing? Or is an error during processing considered fatal?

Reply:
fixed